etna/transforms/math/lags.py

```diff
@@ -96,3 +98,148 @@ def _transform(self, df: pd.DataFrame) -> pd.DataFrame:
+
+        Returns
+        -------
+        result: ExogShiftTransform
+        """
+        feature_names = self._get_feature_names(df=df)
+
+        self._exog_shifts = dict()
+        self._created_regressors = []
+
+        for feature_name in feature_names:
+            shift = self._estimate_shift(df=df, feature_name=feature_name)
+            self._exog_shifts[feature_name] = shift
+
+            if shift > 0:
+                self._created_regressors.append(f"{feature_name}_shift_{shift}")
+
+        return self
+
+    @staticmethod
+    def _get_feature_names(df: pd.DataFrame) -> List[str]:
+        """Return the names of exogenous variables."""
+        names = set(df.columns.get_level_values("feature"))
+        if "target" in names:
+            names.remove("target")
+
+        return list(names)
+
+    def _estimate_shift(self, df: pd.DataFrame, feature_name: str) -> int:
+        """Estimate shift value for exogenous variable."""
+        if not self.auto:
+            return self.lag  # type: ignore
+
+        freq = pd.infer_freq(df.index)
+
+        last_date = df.index.max() + pd.Timedelta(self.horizon, unit=freq)
+        last_feature_date = self._exog_last_date[feature_name]  # type: ignore
+
+        delta = last_date - last_feature_date
+        shift = max(0, int(delta / pd.Timedelta(1, unit=freq)))
+
+        return shift
+
+    def _transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Shift exogenous variables.
+
+        Parameters
+        ----------
+        df:
+            dataframe with data to transform.
+
+        Returns
+        -------
+        result: pd.Dataframe
+            transformed dataframe
+        """
+        if self._exog_shifts is None:
+            raise ValueError("Transform is not fitted!")
+
+        result = df
+        freq = pd.infer_freq(df.index)
+        segments = sorted(set(df.columns.get_level_values("segment")))
+        feature_names = self._get_feature_names(df=df)
+
+        shifted_features = []
+        features_to_remove = []
+        for feature_name in feature_names:
+            shift = self._exog_shifts[feature_name]
+
+            feature = df.loc[:, pd.IndexSlice[:, feature_name]]
+
+            if shift > 0:
+                shifted_feature = feature.shift(shift, freq=freq)
+
+                column_name = f"{feature_name}_shift_{shift}"
+                shifted_feature.columns = pd.MultiIndex.from_product([segments, [column_name]])
+
+                shifted_features.append(shifted_feature)
+                features_to_remove.append(feature_name)
+
+        if len(features_to_remove) > 0:
+            result = result.drop(columns=pd.MultiIndex.from_product([segments, features_to_remove]))
+
+        result = pd.concat([result] + shifted_features, axis=1)
+        result.sort_index(axis=1, inplace=True)
+        return result
+
+    def get_regressors_info(self) -> List[str]:
+        """Return the list with regressors created by the transform."""
```

Comment:
We should probably check if it is ok to return `[]` if the models isn't fit yet. Because the list of regressors will be known only after fitting.

Reply:
Changed it to raise an error if the transform is not fitted.

Reply:
Do we have this behavior in any other transform?

Reply:
Yes, in `LabelEncoderTransform`

Reply:
Ok, it seems like in `LabelEncoderTransform` and `OneHotEncoderTransform` we have this kind of behavior.